### PR TITLE
[Challenge 2] Implement leader/follower event-loops

### PR DESCRIPTION
**Pull request by khollbach**
(untested!)

The code compiles, but besides that, I have no idea what'll happen when we run it.

Let me know what you think; we can look at testing this next time we're coworking.

### CODE REVIEW

File: src/challenge-2/challenge-2.c [hunk 1]
```diff
@@ -22,10 +29,92 @@ int main(void)
...
+    json_object* node_id = json_object_object_get(body, "node_id");
+    return json_object_get_string(node_id);
+}
+
+// Borrows msg.
+// The returned string is borrowed from msg; don't use it after freeing msg.
+const char* msg_type(json_object* msg) {
+    json_object* body = json_object_object_get(msg, "body");
+    json_object* type = json_object_object_get(body, "type");
+    return json_object_get_string(type);
```
Comment (Per48edjes) [inline src/challenge-2/challenge-2.c]: Not a big deal at all at this stage, but perhaps this makes sense to live in `utils.c` (or some library module; thinking we'll probably be inspecting these fields frequently going forward)? Good food for thought for the conch messaging functions from #3 , too.

File: src/challenge-2/challenge-2.c [hunk 2]
```diff
@@ -22,10 +29,92 @@ int main(void)
...
+            fprintf(stderr, "Error: follower_event_loop: unrecognized message type: \"%s\"", type);
+            free(msg);
+            exit(EXIT_FAILURE);
+        }
+    }
+}
+
+void leader_event_loop() {
+    Queue* queue = queue_init(100);
+    Conch* conch = conch_init(0);
```
Comment (Per48edjes) [inline src/challenge-2/challenge-2.c:62]: 🐚  lol

**Review by Per48edjes — APPROVED** [review]
Took a cursory look...and I liked what I saw! Solid work -- it *looks* elegant.

The next level's Boss is the compiler, which we can tackle when we pair next!

File: src/challenge-2/challenge-2.c (hunk 1 above)
Comment (khollbach) [reply src/challenge-2/challenge-2.c]: Great idea; done!

File: src/challenge-2/challenge-2.c (hunk 2 above)
Comment (khollbach) [reply src/challenge-2/challenge-2.c:62]: :shell: :shell: :shell:

**Pull request merged**